Feature/Member

myu7769 opened the pull request:
### 변경사항
**AS-IS**
Member 로그인 후 Token 생성
SecurityFilter 구현

**TO-BE**
Member 조회/탈퇴 기능 구현
상품 등록/수정/삭제 기능 구현

### 테스트
- [ ] 테스트 코드
- [x] API 테스트

Commit:
feat : Entity 추가

Member, Notification, Photo, Review, Trade 테이블 정의

Commit:
feat : repository 추가

Member, Notification, Photo, Review, Trade repository

myu7769 committed:
feat : Entity 생성 및 수정
feat : SignUp/SignIn api 구현
feat : CustomException 구현
test : MemberService test 코드 작성

Commit:
fix : Entity 수정 및 status enum 생성
fix : MemberService -> SignIn return member 수정
fix : SignController -> url 수정

Commit:
feat : password 암호화 DB 저장
chore : swagger2 추가

Commit:
chore : 패키지 이름 수정

Commit:
chore : 패키지 이름 수정

Commit:
chore : 레디스 설정
fix : Email, Phone 중복 체크

Commit:
chore : 레디스 설정
fix : Email, Phone 중복 체크

Commit:
feat : 로그인 시 jwt 토큰 생성 후 전송

Commit:
feat : 로그인 시 jwt 토큰 생성 후 전송

Commit:
feat : spring security 설정
feat : Member Entity Role 컬럼 추가

Commit:
feat : spring security 설정
feat : Member Entity Role 컬럼 추가

Commit:
docs : erd 수정

Commit:
docs : erd 수정

Commit:
fix : GetRequest는 모두 허용되게 수정

Commit:
fix : GetRequest는 모두 허용되게 수정

Ameri-Kano commented on src/main/java/com/zerobase/trade/domain/entity/Member.java:
```diff
@@ -50,10 +68,42 @@ public static Member of(MemberSignUpForm form, String password) {
                 .password(password)
                 .name(form.getName())
                 .phone(form.getPhone())
+                .roles(Collections.singletonList("USER"))
```
이 부분에서 `List.of()` 가 아닌 `Collections.singletonList()` 를 특별히 활용하는 이유가 있나요?

myu7769 replied:
안녕하세요. Collections.singletonList을 쓴 이유는 현재 제 구현 상 Member의 롤이 하나뿐이고, 하나의 List를 가질 경우엔 singletonList 사용시 메모리 효율성이 있다고 하여 사용하였습니다. 좀 더 찾아보니 값을 변경할 수도 없다고 하네요.

Ameri-Kano commented on src/main/java/com/zerobase/trade/security/token/JwtAuthenticationProvider.java:
```diff
@@ -4,19 +4,42 @@
     private long tokenValidTime = 1000L * 60 * 60 * 24;
 
-    public String createToken(String account, Long id) {
+    // bean 생성 되기 전에 secretKey 인코딩
+    @PostConstruct
+    protected void init() {
+        secretKey = Base64.getEncoder().encodeToString(secretKey.getBytes(StandardCharsets.UTF_8));
+    }
```
Bean 이 생성되는 부분은 어느 부분이고, 생성되기 전에 Secret Key를 인코딩해야 하는 이유가 궁금합니다!

myu7769 replied:
주석으로 Bean이 생성되기 전에 라고 적어두었는데, 빈이 초기화(기본 생성자 실행) 후 한번 실행할 수 있는 어노테이션입니다. 인코딩을 한 이유는 현재 해시를 수행하는 secretKey 값이 너무 단순하여 인코딩했습니다.

Ameri-Kano replied:
답변해주셔서 감사합니다! 몇 가지 몰랐던 걸 배워가네요!

Ameri-Kano reviewed (approved):
코드를 보고 궁금한 부분에 대한 질문을 조금 남겨봤습니다. **(틀린 부분을 고쳐달라는 요청이 아닙니다!)** 질문에 답변 해주셔도 되고 안해주셔도 괜찮습니다!
감사합니다!

mingming-mentor commented on src/main/java/com/zerobase/trade/config/RedisConfig.java:
```diff
@@ -0,0 +1,45 @@
+
+  @Bean
+  public RedisTemplate<String, Member> redisTemplate(){
+    RedisTemplate<String, Member> redisTemplate = new RedisTemplate<>();
```
entity를 그대로 캐싱하면 안될것 같네요. dto로 변환해서 사용해주세요..! entity를 그대로 caching하면 redis cache에서 get 하였을때 객체 영속성에 들어온 entity가 아님에도 착각하여 사용할 수 있게 될듯 합니다.

myu7769 replied:
네.. 그럴 수도 있겠군요. caching 된 값과 Entity 자체를 구분하도록 DTO 만들어 구현하겠습니다.

mingming-mentor commented on src/main/java/com/zerobase/trade/config/RedisConfig.java:
```diff
@@ -0,0 +1,45 @@
+    RedisTemplate<String, Member> redisTemplate = new RedisTemplate<>();
+    redisTemplate.setConnectionFactory(redisConnectionFactory());
+    redisTemplate.setKeySerializer(new StringRedisSerializer());
+    redisTemplate.setValueSerializer(new StringRedisSerializer());
```
value는 string 타입이 아닌데 stringRedisSerializer를 이용하나요?

myu7769 replied:
value 값은 객체니까 GenericJackson2JsonRedisSerializer 사용해 구현하겠습니다.

mingming-mentor commented on src/main/java/com/zerobase/trade/domain/entity/Member.java:
```diff
@@ -19,26 +26,37 @@
 @AllArgsConstructor
 @Table(name = "member")
-public class Member{
+public class Member implements UserDetails {
```
UserDetails도 entity에서 구현하기보단, dto에서 구현해주시면 좋을 것 같아요. entity는 table을 java 코드로 맵핑한 것으로 그 외의 부가적인 기능이 있다면 가독성 측면에서 떨어질 것 같습니다.

myu7769 replied:
@mingming-mentor 님 리뷰 감사합니다. DTO 사용해 구현하겠습니다. 위 리뷰 중 Redis에서 사용되는 DTO와 security 에서 사용되는 DTO를 하나로 만드는 게 좋을가요? 실무에서는 어떻게 사용되는 지 궁금합니다.

mingming-mentor replied:
@myu7769 님, 해당 부분은 하나로 사용하셔도 무방할 듯 합니다!

mingming-mentor reviewed (approved):
수고 많으셨어요! 몇가지 리뷰 드렸습니다

myu7769 commented:
리뷰 감사합니다.

JaeYooooon commented on src/main/resources/application.properties:
Secret Key 는 환경변수로 관리하는건 어떨까요??

myu7769 committed:
feat : member 삭제 추가
fix : Redis 참조 객체를 entity Member에서 memberDTO로 변경
fix : security UserDetails 상속 클래스 memberDTO로 변경

myu7769 committed:
feat : member 삭제 추가
fix : Redis 참조 객체를 entity Member에서 memberDTO로 변경
fix : security UserDetails 상속 클래스 memberDTO로 변경

myu7769 committed:
Merge remote-tracking branch 'origin/Feature/Member' into Feature/Member
